src/edit.rs

```diff
@@ -614,13 +641,17 @@ impl<'out, 'prompt, H: Helper> State<'out, 'prompt, H> {
         ) {
             self.ctx.history_index = history_index;
             let buf = history.get(history_index).unwrap();
-            self.changes.borrow_mut().begin();
-            self.line.update(buf, buf.len());
-            self.changes.borrow_mut().end();
-            self.refresh_line()
+            let pos = if keep_cursor_pos {
+                self.line.pos()
+            } else {
+                buf.len()
+            };
+            self.update_line(buf, pos);
         } else {
-            self.out.beep()
+            self.restore();
```

Comment:
I don't know if this `restore` breaks the following code:

https://github.com/kkawakam/rustyline/blob/88e669c4ff17c1e7cc59f5eeff12ac3dc068c5c6/src/lib.rs#L558-L559